go/chat/signencrypt/codec.go

```diff
@@ -565,6 +566,63 @@ func OpenWhole(sealed []byte, encKey SecretboxKey, verifyKey VerifyKey, signatur
 	return append(output, moreOutput...), nil
 }
 
+// include verification of associated data. see
+// https://en.wikipedia.org/wiki/Authenticated_encryption#Authenticated_encryption_with_associated_data_(AEAD)
+
+type AEADMessage struct {
+	Version       int               `codec:"v" json:"v"`
+	AssocDataHash [sha512.Size]byte `codec:"a" json:"a"`
+	Message       []byte            `codec:"m" json:"m"`
+}
+
+// SealWithAssociatedData is a wrapper around SealWhole which adds an associatedData object
+// (see AEAD ciphers) which must be message-packable into bytes. This exact object is required
+// to call OpenWithAssociatedData on the ciphertext.
+func SealWithAssociatedData(msg []byte, associatedData interface{}, encKey SecretboxKey, signKey SignKey, signaturePrefix kbcrypto.SignaturePrefix, nonce Nonce) (ret []byte, err error) {
+	adEncoded, err := msgpack.Encode(associatedData)
```

Comment:
Chat uses `codec.MsgpackHandle{WriteExt: true}` I wonder if `WriteExt` is relevant

Reply:
just dove into this. it does seem that most of the most similar binary encoding/decoding uses this pattern. good news though. msgpack.Encode and msgpack.Decode are just shorthand for exactly this. :)